protocols/rendezvous/src/client.rs

```diff
@@ -218,159 +222,174 @@ impl NetworkBehaviour for Behaviour {
+    fn on_swarm_event(&mut self, event: FromSwarm<Self::ConnectionHandler>) {
+        self.external_addresses.on_swarm_event(&event);
+
+        self.inner.on_swarm_event(event);
     }
 
     fn poll(
         &mut self,
         cx: &mut Context<'_>,
-        _: &mut impl PollParameters,
+        params: &mut impl PollParameters,
     ) -> Poll<ToSwarm<Self::ToSwarm, THandlerInEvent<Self>>> {
-        if let Some(event) = self.events.pop_front() {
-            return Poll::Ready(event);
+        if let Some(event) = self.error_events.pop_front() {
+            return Poll::Ready(ToSwarm::GenerateEvent(event));
         }
 
-        if let Some((namespace, rendezvous_node, ttl)) = self.pending_register_requests.pop() {
-            // Update our external addresses based on the Swarm's current knowledge.
-            // It doesn't make sense to register addresses on which we are not reachable, hence this should not be configurable from the outside.
-
-            let external_addresses = self.external_addresses.iter().cloned().collect::<Vec<_>>();
+        let result = self.inner.poll(cx, params).map(|to_swarm| {
+            to_swarm.map_out(|event| match event {
+                libp2p_request_response::Event::Message {
+                    peer: _,
+                    message:
+                        libp2p_request_response::Message::Response {
+                            request_id,
+                            response,
+                        },
+                } => self.handle_response(&request_id, response),
+                libp2p_request_response::Event::OutboundFailure {
+                    peer: _,
+                    request_id,
+                    error: _,
+                } => {
+                    let (rendezvous_node, namespace) = self
+                        .waiting_for_register
+                        .remove(&request_id)
+                        .unwrap_or_else(|| panic!("unknown request_id: {request_id}"));
+                    Event::RegisterFailed(RegisterError::Remote {
+                        rendezvous_node,
+                        namespace,
+                        error: ErrorCode::Unavailable,
+                    })
+                }
+                // By specification, a client able only send a request (get a response)
+                // and can't receive any inbound request.
+                // That is why the next cases should be considered as unreachable.
+                libp2p_request_response::Event::InboundFailure { .. } => {
+                    unreachable!()
+                }
+                libp2p_request_response::Event::ResponseSent { .. } => {
+                    unreachable!()
+                }
+                libp2p_request_response::Event::Message {
+                    peer: _,
+                    message: libp2p_request_response::Message::Request { .. },
+                } => {
+                    unreachable!()
+                }
```

Comment:
Thank you for adding this. I'd prefer it to be in the `unreachable!()` macro itself:

```suggestion
                libp2p_request_response::Event::InboundFailure { .. } | libp2p_request_response::Event::ResponseSent { .. } | libp2p_request_response::Event::Message {
                    peer: _,
                    message: libp2p_request_response::Message::Request { .. },
                } => {
                    unreachable!("rendezvous clients never receive requests")
                }
```